### PR TITLE
YC-578: Add token authentication for rest/permits api and filter request by ip

### PR DESCRIPTION
**Pull request by monodo**

### CODE REVIEW

**Review by danickfort-liip — CHANGES_REQUESTED** [review]
@monodo J'ai trouvé un bug et quelques remarques/questionnements concernant cette PR.

File: permits/services.py [hunk 1]
```diff
@@ -1502,15 +1502,24 @@ def clear_session_filters(request):
...
+    return False
+
+
+def check_request_comes_from_internal_qgisserver(request):
+    """
+    Check that the request is coming from inside the docker composition AND that it is an allowed ip
+    """
+    if (
+        check_request_ip_is_in_whithelist(request)
+        and socket.gethostbyname("qgisserver") == request.META["REMOTE_ADDR"]
```
[the remark this answers is not in the document]
Comment (monodo) [reply permits/services.py:1522]: Non, c'est déterminé dans la composition docker

File: permits/services.py [hunk 2]
```diff
@@ -1502,15 +1502,24 @@ def clear_session_filters(request):
...
-    Check that the request is coming from inside the docker composition AND that it is a private IP
+    Check that the request is coming from allowed ip
     """
-
-    for whitelisted_ip in settings.LOCAL_IP_WHITELIST:
-        if (
-            request.META["REMOTE_ADDR"].startswith(whitelisted_ip)
-            and socket.gethostbyname("qgisserver") == request.META["REMOTE_ADDR"]
-        ):
+    for whitelisted_ip in config.IP_WHITELIST:
```
[the remark this answers is not in the document]
Comment (monodo) [reply permits/services.py]: Done

[the remark this answers is not in the document]
Comment (monodo) [reply permits/services.py]: Ok pour toi @danickfort-liip https://github.com/yverdon/geocity/blob/yc-578/permits/tests/test_permit_request_api.py#L306-L318 ?

[the remark this answers is not in the document]
Comment (danickfort-liip) [reply permits/services.py]: OK 👍

**Review by danickfort-liip — APPROVED** [review]

**Comment by rbovard** [comment]
Thanks a lot! It's working fine on my side, with defining ip in `IP_WHITELIST` or range in `NETWORK_WHITELIST`.

Question: should we keep the default values? It will be more secure if by default nothing is set when enabling `DRF_ALLOW_TOKENAUTHENTICATION`.

File: env.demo [hunk 3]
```diff
@@ -63,3 +62,8 @@ AXES_FAILURE_LIMIT=5
 # Lock out by combination of ip AND User
 AXES_LOCK_OUT_BY_COMBINATION_USER_AND_IP=True
 AXES_COOLOFF_TIME=2
+#Limit requests on DRF. Note that this might impact the print service!
+DRF_THROTTLE_RATE_PERMITS_API='1000/day'
+DRF_THROTTLE_RATE_EVENTS_API='100000/day'
+#This allow access to the API with TokenAuthentication
+DRF_ALLOW_TOKENAUTHENTICATION=False
```
Comment (rbovard) [inline env.demo:69]: Missing new line

File: permits/services.py [hunk 4]
```diff
@@ -1502,15 +1503,44 @@ def clear_session_filters(request):
...
 
-    for whitelisted_ip in settings.LOCAL_IP_WHITELIST:
-        if (
-            request.META["REMOTE_ADDR"].startswith(whitelisted_ip)
-            and socket.gethostbyname("qgisserver") == request.META["REMOTE_ADDR"]
-        ):
-            return True
+    if (
+        check_request_ip_is_allowed(request)
+        and socket.gethostbyname("qgisserver") == request.META["REMOTE_ADDR"]
```
Comment (rbovard) [inline permits/services.py:1543]: Need to use `get_client_ip()`?

File: geomapshark/settings.py [hunk 5]
```diff
@@ -271,6 +272,16 @@
...
     "TITLE_COLOR": ("#000000", "Couleur du titre", str,),
     "TABLE_COLOR": ("#212529", "Couleur du texte dans les tableaux", str,),
+    "IP_WHITELIST": (
+        "172.16.0.1,172.17.0.1,127.0.0.1,localhost",
+        "IP  autorisées pour l'utilisation de l'API complète (/rest/permits), séparées par des ','",
+        str,
+    ),
+    "NETWORK_WHITELIST": (
+        "172.16.0.0/12,192.168.0.0/16",
+        "Réseaux  autorisés pour l'utilisation de l'API complète (/rest/permits), séparées par des ','",
```
Comment (rbovard) [inline geomapshark/settings.py:282]:
```suggestion
        "Réseaux autorisés pour l'utilisation de l'API complète (/rest/permits), séparés par des ','",
```

File: geomapshark/settings.py [hunk 6]
```diff
@@ -271,6 +272,16 @@
     "TEXT_COLOR": ("#000000", "Couleur du texte", str,),
     "TITLE_COLOR": ("#000000", "Couleur du titre", str,),
     "TABLE_COLOR": ("#212529", "Couleur du texte dans les tableaux", str,),
+    "IP_WHITELIST": (
+        "172.16.0.1,172.17.0.1,127.0.0.1,localhost",
+        "IP  autorisées pour l'utilisation de l'API complète (/rest/permits), séparées par des ','",
```
Comment (rbovard) [inline geomapshark/settings.py:277]:
```suggestion
        "IP autorisées pour l'utilisation de l'API complète (/rest/permits), séparées par des ','",
```

**Review by rbovard — CHANGES_REQUESTED** [review]

**Review by rbovard — APPROVED** [review]

**Pull request merged**